### ISSUE TITLE
Broken deployment of assets - assets file is not updated on deployment

### ISSUE THREAD

**Issue by luisramos0** [openfoodfoundation/ofn-install#633]
After provisioning and deploy of ES staging.

The website renders like this:
![image](https://user-images.githubusercontent.com/1640378/83555898-eb0b6200-a506-11ea-8ed1-17b398fe729f.png)

The console looks like this:
![image](https://user-images.githubusercontent.com/1640378/83555938-f8c0e780-a506-11ea-8d88-62a0fcb719b5.png)

Maybe not related but these are some of the most common errors in bugsnag:
![image](https://user-images.githubusercontent.com/1640378/83556101-3b82bf80-a507-11ea-99f7-98327293ea6a.png)

The javascript tag is:
![image](https://user-images.githubusercontent.com/1640378/83555878-e34bbd80-a506-11ea-94e0-4059c182a4b7.png)

But in /home/openfoodnetwork/apps/openfoodnetwork the server has:
current/public/assets/darkswarm/all-0b7561740168a5238cd54ae46030d6d6.js
The linked bundle is present in an previous release (it's actually two releases ago in staging ES, not the previous one):
releases-old/2020-06-02-081616/public/assets/darkswarm/all-7afe1299726a714354d01b103ef4fed0.js

**Comment by luisramos0** [comment]
The problem can be checked on the server with these commands:

```
openfoodnetwork@staging:~/apps/openfoodnetwork/current$ ls -lart public/assets/darkswarm/all-*js
-rw-rw-r-- 1 openfoodnetwork openfoodnetwork 2843051 jun  2 17:33 public/assets/darkswarm/all-0b7561740168a5238cd54ae46030d6d6.js
openfoodnetwork@staging:~/apps/openfoodnetwork/current$ curl https://staging.katuma.org/ | grep darkswarm | grep js
<script src="/assets/darkswarm/all-7afe1299726a714354d01b103ef4fed0.js" type="text/javascript"></script>
```

**Comment by sauloperez** [comment]
I manually precompiled the assets and restarted the server and they are back. We'll continue investigating because https://github.com/openfoodfoundation/ofn-install/pull/629 was supposed to fix this.

**Comment by sauloperez** [comment]
I found https://stackoverflow.com/a/14720284/2838228 while investigating the issue and I think it's pretty useful. Something else to add to our toolbelts :ok_hand:

**Comment by sauloperez** [comment]
something is pretty broken. In AU the spree/taxons images can't be found as well as some system/content_configurations :thinking_face: , while in UK simply the digests don't match

```
<script src="/assets/darkswarm/all-425fb4c567a27acb28fb10577cd824bb.js"></script>
```

```
irb(main):004:0> Rails.application.assets.find_asset('darkswarm/all.js')
=> #<Sprockets::BundledAsset:0x2afa023f5e68 pathname="/home/openfoodnetwork/apps/openfoodnetwork/releases-old/2020-06-03-113502/app/assets/javascripts/darkswarm/all.js.coffee", mtime=2020-06-03 11:36:19 +0000, digest="ebc7f9755b9b49f9ca483c89984d2b82">
```

**Comment by Matt-Yorkley** [comment]
With regards to https://stackoverflow.com/a/14720284/2838228 maybe can we just add these methods into the two places where we actually include the JS and CSS files in the DOM?

**Comment by Matt-Yorkley** [comment]
> In AU the spree/taxons images can't be found as well as some system/content_configurations

I think some image assets can get fingerprinted as well, depending on what path they are in.

**Comment by luisramos0** [comment]
"I manually precompiled the assets and restarted the server and they are back. "

@sauloperez I did that in staging ES yesterday after provisioning and deploy with latest ofn-install and it didnt fix the problem :man_shrugging:

**Comment by sauloperez** [comment]
I think what really does the trick is restarting @luisramos0 but the ordering of the actions might matter. I can't find a pattern either though.

**Comment by luisramos0** [comment]
wait, restarting unicorn? I did that.

**Comment by sauloperez** [comment]
I just did it with UK staging. Doing a `sudo systemctl reload unicorn_openfoodnetwork` had no effect, while a `sudo systemctl restart unicorn_openfoodnetwork` did.

**Comment by sauloperez** [comment]
I also spotted something that doesn't seem normal. Is 

```
~/apps/openfoodnetwork/releases-old$ ll 2020-06-01-190250/public/assets/darkswarm/all-*.js | wc -l
91
```

All of these with different digests, While the latest release lists only

```
~/apps/openfoodnetwork/releases-old$ ll 2020-06-03-113502/public/assets/darkswarm/
total 6336
drwxr-xr-x  2 openfoodnetwork openfoodnetwork    4096 Jun  3 11:38 ./
drwxr-xr-x 20 openfoodnetwork openfoodnetwork   12288 Jun  3 11:38 ../
-rw-r--r--  1 openfoodnetwork openfoodnetwork  402907 Jun  3 11:35 all.css
-rw-r--r--  1 openfoodnetwork openfoodnetwork   46107 Jun  3 11:38 all.css.gz
-rw-r--r--  1 openfoodnetwork openfoodnetwork 2217577 Jun  3 11:36 all-ebc7f9755b9b49f9ca483c89984d2b82.js
-rw-r--r--  1 openfoodnetwork openfoodnetwork  558752 Jun  3 11:36 all-ebc7f9755b9b49f9ca483c89984d2b82.js.gz
-rw-r--r--  1 openfoodnetwork openfoodnetwork  403006 Jun  3 11:35 all-faa5b805bc1495719155e73d45854602.css
-rw-r--r--  1 openfoodnetwork openfoodnetwork   46195 Jun  3 11:35 all-faa5b805bc1495719155e73d45854602.css.gz
-rw-r--r--  1 openfoodnetwork openfoodnetwork 2217577 Jun  3 11:36 all.js
-rw-r--r--  1 openfoodnetwork openfoodnetwork  558752 Jun  3 11:38 all.js.gz
```

**Comment by Matt-Yorkley** [comment]
The digests are just an md5 of the file contents, right?

**Comment by sauloperez** [comment]
Yep. The strange thing is that we have 91 copies of the `darkswarm/all.js` within the same release.

**Comment by luisramos0** [comment]
ok Pau, yesterday on staging ES I tried reload and then restart. no effect, could be related to unicorn workers, I am not sure what is the exact process where unicorn picks up the fingerpint value.
I see those 90+ files but in the latest releases the files seem correct, it's the fingerprint on the html produced by unicorn that is outdated....

**Comment by Matt-Yorkley** [comment]
> right after reloading a third strange process showed up
```
openfoodnetwork@staging:~/apps/openfoodnetwork/releases-old$ ps axuf | grep unic
openfoo+ 19530  0.0  2.4 482032 198720 ?       Sl   Jun02   0:11 unicorn master (old) -D -c /home/openfoodnetwork/apps/openfoodnetwork/shared/config/unicorn.rb -E staging
openfoo+ 19594  0.0  4.5 993732 371652 ?       Sl   Jun02   1:18  \_ unicorn worker[0] -D -c /home/openfoodnetwork/apps/openfoodnetwork/shared/config/unicorn.rb -E staging
openfoo+ 19597  0.1  4.8 1014980 394972 ?      Sl   Jun02   1:31  \_ unicorn worker[1] -D -c /home/openfoodnetwork/apps/openfoodnetwork/shared/config/unicorn.rb -E stagingopenfoo+  3640  101  0.5 212280 48984 ?        Rl   14:44   0:02  \_ ruby /home/openfoodnetwork/.gem/ruby/2.3.0/bin/unicorn -D -c /home/openfoodnetwork/apps/openfoodnetwork/shared/config/unicorn.rb -E staging
```

How long did you watch it for? The new method will mark the current master process as "old", then spawn a new master process, then when that one starts creating new workers it will shut down the old master process. It takes maybe 15 seconds, depending on what the workers are doing.

**Comment by sauloperez** [comment]
That's exactly what I wanted to know. It was right after so I'm pretty sure the replacement was still taking place. I'm deploying again to see if things keep consistently failing this way or it was a matter of a second deploy after provisioning :sweat_smile:

**Comment by sauloperez** [comment]
I can confirm precompiling and unicorn reloading do work. Subsequent deploys do not change the digest on `public/assets/darkswarm/all-*.js` and assets load correctly. Then, changing `public/assets/admin/all-*.js` in https://github.com/openfoodfoundation/openfoodnetwork/pull/5554 lead to a different digest compared to the previous release.

The only staging having issues is AU, but only with `system/content_configurations/`, not loading the home page hero image, for instance. AFAIK these are assets set from the instance config, which makes me think this is a) a separate issue b) likely preexisting. Can you confirm @filipefurtad0 ?

**Comment by filipefurtad0** [comment]
Hi @sauloperez,

I can confirm that as of now (i.e., didn't deploy to any of the servers above yet) only staging-AU shows rendering issues on the /admin page. I'll resume staging/testing soon and will drop a line if I notice anything unusual.

Thanks!

**Comment by filipefurtad0** [comment]
Hi @sauloperez,

same issue, after staging #5414 on katuma staging-server:
![image](https://user-images.githubusercontent.com/49817236/83693457-04d3a480-a5ee-11ea-9e5c-06264dd1f0b4.png)

This is tricky: it's only observable in private browsing or after cleaning cookies, which makes it look fine if one just opens a new tab, or refreshes page, after staging.

I didn't observe this or any issues after staging/testing #5425 in FR earlier today: the change the PR introduced was visible, which wasn't the case when it was staged on staging-es before...

**Comment by luisramos0** [comment]
I deployed several times with master and the problem remained.
Deploying with #635 fixed the problem on both ES and FR.
I think we should go ahead with #635 and then see what's wrong with the zero dowtime process.

**Comment by Matt-Yorkley** [comment]
I can consistently trigger this bug now by deploying PRs that have different gems, eg `master` followed by a dependabot PR. The unicorn logs show this error:

```
I, [2020-06-04T15:58:50.876913 #1624]  INFO -- : executing ["/home/openfoodnetwork/.gem/ruby/2.3.0/bin/unicorn", "-D", "-c", "/home/openfoodnetwork/apps/openfoodnetwork/shared/config/unicorn.rb", "-E", "staging", {15=>#<Kgio::UNIXServer:fd 15>}] (in /home/openfoodnetwork/apps/openfoodnetwork/releases-old/2020-06-04-155407)
/home/openfoodnetwork/.rbenv/versions/2.3.7/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require': cannot load such file -- bundler/setup (LoadError)
	from /home/openfoodnetwork/.rbenv/versions/2.3.7/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
```

There's some issue between `bundler` and `unicorn` at the point where the master processes are changing over.

I think for now we can go with this: https://github.com/openfoodfoundation/ofn-install/issues/638. Using a restart instead of reload seems to resolve the issue, so it should unblock these staging/testing problems for now.

**Comment by sauloperez** [comment]
Can you please create an issue with this insights @Matt-Yorkley or have you alread? we'll need to retake the 0-downtime deployments from there.
